services/requester/pool.go

```diff
@@ -102,6 +154,224 @@ func (t *TxPool) Send(
+// This helps to reduce the nonce mismatch errors which mainly occur from the
+// re-ordering of Cadence transactions that happens from Collection nodes.
+type BatchTxPool struct {
+	*SingleTxPool
+	pooledTxs map[gethCommon.Address][]pooledEvmTx
+	txMux     sync.Mutex
+}
+
+var _ TxPool = &BatchTxPool{}
+
+func NewBatchTxPool(
+	client *CrossSporkClient,
+	transactionsPublisher *models.Publisher[*gethTypes.Transaction],
+	logger zerolog.Logger,
+	config config.Config,
+	collector metrics.Collector,
+	keystore *keystore.KeyStore,
+) *BatchTxPool {
+	// initialize the available keys metric since it is only updated when sending a tx
+	collector.AvailableSigningKeys(keystore.AvailableKeys())
+
+	singleTxPool := NewSingleTxPool(
+		client,
+		transactionsPublisher,
+		logger,
+		config,
+		collector,
+		keystore,
+	)
+	batchPool := &BatchTxPool{
+		SingleTxPool: singleTxPool,
+		pooledTxs:    make(map[gethCommon.Address][]pooledEvmTx),
+		txMux:        sync.Mutex{},
+	}
+
+	go batchPool.processPooledTransactions()
+
+	return batchPool
+}
+
+// Add adds the EVM transaction to the tx pool, grouped with the rest of the
+// transactions from the same EOA signer.
+// After the configured `TxBatchInterval`, the collected transations
+// are batched and sent to the Flow network using `EVM.batchRun`, for execution.
+func (t *BatchTxPool) Add(
+	ctx context.Context,
+	tx *gethTypes.Transaction,
+) error {
+	// tx sending should be blocking, so we don't have races when
+	// pooled transactions are being processed in the background.
+	t.txMux.Lock()
+	defer t.txMux.Unlock()
+
+	t.txPublisher.Publish(tx) // publish pending transaction event
+
+	from, err := gethTypes.Sender(gethTypes.LatestSignerForChainID(tx.ChainId()), tx)
+	if err != nil {
+		return fmt.Errorf("failed to derive the sender: %w", err)
+	}
+	txData, err := tx.MarshalBinary()
+	if err != nil {
+		return err
+	}
+	hexEncodedTx, err := cadence.NewString(hex.EncodeToString(txData))
+	if err != nil {
+		return err
+	}
+
+	userTx := pooledEvmTx{txPayload: hexEncodedTx, nonce: tx.Nonce()}
+	t.pooledTxs[from] = append(t.pooledTxs[from], userTx)
+
+	return nil
+}
+
+func (t *BatchTxPool) processPooledTransactions() {
+	for range time.Tick(t.config.TxBatchInterval) {
+		// Take a snapshot here to allow `Add()` continue accepted
+		// incoming EVM transactions, without blocking until the
+		// batch transactions are submitted.
+		snapshot := func() map[gethCommon.Address][]pooledEvmTx {
+			t.txMux.Lock()
+			defer t.txMux.Unlock()
+			copy := t.pooledTxs
+			t.pooledTxs = make(map[gethCommon.Address][]pooledEvmTx)
+			return copy
+		}()
+
+		for address, pooledTxs := range snapshot {
+			if err := t.batchSubmitTransactions(pooledTxs); err != nil {
+				t.logger.Error().Err(err).Msgf(
+					"failed to send Flow transaction from BatchPool for EOA: %s",
+					address.Hex(),
+				)
+				continue
+			}
+		}
+	}
+}
+
+func (t *BatchTxPool) batchSubmitTransactions(
+	pooledTxs []pooledEvmTx,
```

Comment:
maybe we could add the address for the txs, so that the address can be useful to be included in the error message?

Reply:
We already do that in the caller of this function:
```go
if err := t.batchSubmitTransactionsForSameAddress(pooledTxs); err != nil {
	t.logger.Error().Err(err).Msgf(
		"failed to send Flow transaction from BatchPool for EOA: %s",
		address.Hex(),
		)
	continue
}
```
The EOA address is included in the error message.